packages/piro/src/Piro_PerformAnalysis.cpp

```diff
@@ -748,6 +758,340 @@ if(useCustomDotProduct) {
+
+  constr.setSolveParameters(rolParams.sublist("ROL Options"));
+  constr.setNumResponses(piroTSolver->num_g());
+
+  ROL::Ptr<ROL::DynamicObjective<double> > obj_ptr = ROL::makePtrFromRef(obj);
+  ROL::Ptr<ROL::DynamicConstraint<double> > constr_ptr = ROL::makePtrFromRef(constr);
+
+  ROL::Ptr<ROL::Vector<double> > rol_p_ptr = ROL::makePtrFromRef(rol_p);
+  ROL::Ptr<ROL::Vector<double> > rol_x_ptr = ROL::makePtrFromRef(rol_x);
+  ROL::Ptr<ROL::Vector<double> > rol_lambda_ptr = ROL::makePtrFromRef(rol_lambda);
+
+  int seed = rolParams.get<int>("Seed For Thyra Randomize", 42);
+
+  //! set initial guess (or use the one provided by the Model Evaluator)
+  std::string init_guess_type = rolParams.get<string>("Parameter Initial Guess Type", "From Model Evaluator");
+  if(init_guess_type == "Uniform Vector")
+    rol_p.putScalar(rolParams.get<double>("Uniform Parameter Guess", 1.0));
+  else if(init_guess_type == "Random Vector") {
+    Teuchos::Array<double> minmax(2); minmax[0] = -1; minmax[1] = 1;
+    minmax = rolParams.get<Teuchos::Array<double> >("Min And Max Of Random Parameter Guess", minmax);
+    ::Thyra::randomize<double>( minmax[0], minmax[1], rol_p.getVector().ptr());
+  }
+  else if(init_guess_type != "From Model Evaluator") {
+    TEUCHOS_TEST_FOR_EXCEPTION(true, Teuchos::Exceptions::InvalidParameter,
+              std::endl << "Piro::PerformSROLAnalysis, ERROR: " <<
+              "Parameter Initial Guess Type \"" << init_guess_type << "\" is not Known.\nValid options are: \"Parameter Scalar Guess\", \"Uniform Vector\" and \"Random Vector\""<<std::endl);
+  }
+
+  bool useFullSpace = rolParams.get("Full Space",false);
+
+  if(analysisVerbosity >= 3) {
+    *out << "\nPiro PerformAnalysis: ROL options:" << std::endl;
+    rolParams.sublist("ROL Options").print(*out);
+    *out << std::endl;
+  }
+
+  Teuchos::RCP<ROL::BoundConstraint<double> > boundConstraint;
+  bool boundConstrained = rolParams.get<bool>("Bound Constrained", false);
+
+  if(boundConstrained) {
+    Teuchos::RCP<Thyra::VectorBase<double>> p_lo = model->getLowerBounds().get_p(0)->clone_v();
+    Teuchos::RCP<Thyra::VectorBase<double>> p_up = model->getUpperBounds().get_p(0)->clone_v();
+
+    //ROL::Thyra_BoundConstraint<double> boundConstraint(p_lo->clone_v(), p_up->clone_v(), eps_bound);
+    boundConstraint = rcp( new ROL::Bounds<double>(ROL::makePtr<ROL::ThyraVector<double> >(p_lo), ROL::makePtr<ROL::ThyraVector<double> >(p_up)));
+  }
+
+  int return_status = 0;
+
+  RolOutputBuffer<char> rolOutputBuffer;
+  std::ostream rolOutputStream(&rolOutputBuffer);
+  Teuchos::RCP<Teuchos::FancyOStream> rolOutput = Teuchos::getFancyOStream(Teuchos::rcpFromRef(rolOutputStream));
+  rolOutput->setOutputToRootOnly(0);
+
+  Teuchos::RCP<Thyra::VectorBase<double> > scaling_vector_p = p->clone_v();
+  ::Thyra::put_scalar<double>( 1.0, scaling_vector_p.ptr());
+  ROL::PrimalScaledThyraVector<double> rol_p_primal(p, scaling_vector_p);
+
+  if ( useFullSpace ) {
```

Comment:
If the fullSpace code is not tested or if it is not working then we should error out when full space is selected.